File: 1169-42-2127-maximum-employees-to-be-invited-to-a-meeting/1169-42-2127-maximum-employees-to-be-invited-to-a-meeting.cpp
class Solution {
public:
    int maximumInvitations(vector<int>& favorite) {
        int n = favorite.size();
        vector<int>deg(n);
        for(auto it : favorite)deg[it]++;
        queue<int>q;
        for(int i = 0 ; i < n;i++)
            if(!deg[i])q.push(i);
        vector<int>chain(n,1);

        while(q.size()){
            int x = q.front();q.pop();
            int y = favorite[x];
            chain[y] = max(chain[y],chain[x]+1);
            deg[y]--;
            if(!deg[y])q.push(y);
        }

        int mx = 0, meetings = 0;
        for(int i = 0 ; i < n;i++){
            if(!deg[i])continue;
            int cur = i , len = 0;
            while(deg[cur]){
                len++;
                deg[cur]=0;
                cur = favorite[cur];
            }
            if(len > 2)mx=max(mx,len);
            else meetings+=chain[i] + chain[favorite[i]];
        }
        return max(mx,meetings);


    }
};

File: 1169-42-2127-maximum-employees-to-be-invited-to-a-meeting/1169-42-2127-maximum-employees-to-be-invited-to-a-meeting.py
from collections import deque

class Solution:
    def maximumInvitations(self, favorite: list[int]) -> int:
        n = len(favorite)
        deg = [0] * n
        for fav in favorite:
            deg[fav] += 1

        q = deque(i for i in range(n) if deg[i] == 0)
        chain = [1] * n

        while q:
            x = q.popleft()
            y = favorite[x]
            chain[y] = max(chain[y], chain[x] + 1)
            deg[y] -= 1
            if deg[y] == 0:
                q.append(y)

        mx = meetings = 0
        for i in range(n):
            if deg[i] == 0:
                continue
            cur, length = i, 0
            while deg[cur]:
                length += 1
                deg[cur] = 0
                cur = favorite[cur]
            if length > 2:
                mx = max(mx, length)
            else:
                meetings += chain[i] + chain[favorite[i]]

        return max(mx, meetings)
